### PR TITLE
Adds torch.testing.fixtures, TestCase.compare_with_numpy

### PR DESCRIPTION
**Pull request by mruberry**
This PR adds torch.testing.fixtures, a test helper function for generating interesting input tensors. The eventual goal of this function is to replace most ad-hoc and existing tensor generation methods, like _make_tensor and _make_tensors in test_torch.py, although it's currently missing some necessary functionality to do so. test_isfinite_isinf_isnan in test_torch is updated to make use of the new function and demonstrate its utility. 

The PR also takes _np_compare, an increasingly popular helper, improves upon it and makes it suitable for use in more tests. This should facilitate OSS PRs that compare with NumPy.

### CODE REVIEW

File: torch/testing/__init__.py [hunk 1]
```diff
@@ -15,6 +16,125 @@
...
+        large_t = v
+    else:
+        large_t = tensor_factory((_nelems_large,), device=device, dtype=dtype)
+        large_t[0:len(vals)] = v
+        results.append(large_t)
+
+    # Creates scalar tensors
+    scalars = []
+    if len(vals) == 0:
+        scalars.append(tensor_factory((1,), device=device, dtype=dtype))
```
Comment (mruberry) [inline torch/testing/__init__.py:77]: This needs to be reshaped to actually be a scalar tensor (currently 1-dim).

File: torch/testing/__init__.py [hunk 2]
```diff
@@ -15,6 +16,140 @@
...
+_int_defaults = [0, 1, 127]
+_bool_defaults = [True, False]
+
+_float_nonfinites = [float('inf'), float('-inf'), float('nan')]
+_float_defaults = [0., -1., 1., -math.pi, math.pi]
+
+_nelems_large = 65
+_nelems_small = 4
+
+# Helper for fixtures. See fixtures documentation.
```
Comment (mruberry) [inline torch/testing/__init__.py:54]: Return "large" tensor, scalar tensor(s), "small" tensor(s)

File: torch/testing/__init__.py [hunk 3]
```diff
@@ -15,6 +16,140 @@
...
+
+    # Creates "large" tensor
+    # NOTE: the large tensor is filled with the values in vals followed by
+    #  random values.
+    if len(vals) > _nelems_large:
+        large_t = v
+    else:
+        large_t = tensor_factory((_nelems_large,), device=device, dtype=dtype)
+        large_t[0:len(vals)].copy_(v)
+        results.append(large_t)
```
Comment (mruberry) [inline torch/testing/__init__.py:78]: <--

File: torch/testing/__init__.py [hunk 4]
```diff
@@ -15,6 +16,140 @@
...
+
+    # Creates scalar tensors
+    scalars = []
+    if len(vals) == 0:
+        scalars.append(tensor_factory((1,), device=device, dtype=dtype))
+    else:
+        for v in vals:
+            scalars.append(torch.tensor((v), device=device, dtype=dtype))
+
+    return results
```
Comment (mruberry) [inline torch/testing/__init__.py:88]: extend with scalars

File: torch/testing/__init__.py [hunk 5]
```diff
@@ -15,6 +16,140 @@
...
+
+# TODO: support division generation by bounding away from zero (min abs)
+# TODO: support nDims parameter (produces square tensors)
+# TODO: support shape parameter
+# TODO: support range parameter (e.g. values between 100 and 200)
+# TODO: support noncontiguous outputs
+# TODO: allow disabling scalar tensor fixtures
```
Comment (mruberry) [inline torch/testing/__init__.py:95]: # TODO: enable disabling small tensor returns, too

File: torch/testing/__init__.py [hunk 6]
```diff
@@ -15,6 +16,140 @@
...
+    non_finites = None
+    defaults = None
+    if dtype.is_floating_point:
+        return _fixtures_helper(vals, device, dtype,
+                                include_nonfinites=include_nonfinites,
+                                include_defaults=include_defaults,
+                                nonfinites=_float_nonfinites,
+                                defaults=_float_defaults,
+                                type_filter=float,
+                                tensor_factory=torch.randn)
```
Comment (mruberry) [inline torch/testing/__init__.py:118]: randn->rand

File: torch/testing/_internal/common_utils.py [hunk 7]
```diff
@@ -881,6 +881,40 @@ def safeCoalesce(self, t):
...
+        np_result = np_fn(a)
+
+        # Coerces NumPy results (where appropriate)
+        if isinstance(torch_result, torch.Tensor) and not isinstance(np_result, np.ndarray):
+            np_result = np.array(np_result)
+        if isinstance(np_result, np.ndarray):
+            np_result = torch.from_numpy(np_result)
+
+        # Copies the NumPy result (if requested)
+        if copy_np_result and isinstance(np_result, np.ndarray):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:913]: try to torch.from_numpy, if failure try to copy and repeat instead of taking a kwarg for this

Comment (mruberry) [reply torch/testing/_internal/common_utils.py:913]: (and move comment near that code)

File: torch/testing/_internal/common_utils.py [hunk 8]
```diff
@@ -881,6 +881,40 @@ def safeCoalesce(self, t):
 
         return tg
 
+    # Compares the given Torch and NumPy functions on the given tensor-like object.
+    #   if copy_np_result is True then NumPy arrays will be copied before being
+    #     compared with the torch tensor. This is useful if the NumPy array has
+    #     negative strides, which would cause the import to PyTorch to fail.
+    # TODO: support bfloat16 comparisons
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:888]: TODO: args/kwargs for torch, numpy, and assertEqual

Comment (mruberry) [reply torch/testing/_internal/common_utils.py:888]: Maybe don't add args/kwargs for torch and numpy functions and suggest pattern of passing lambda (t) -> or partial function fn(t) -> instead
